File: textquest-client/python/textquest/websocket.py
import asyncio
import json
from typing import Optional, Callable, Awaitable, Dict, Any
from enum import Enum

import websockets

from .types import TextQuestError, SessionEvent, SessionEventType


class WebSocketClient:
    def __init__(
        self,
        base_url: str = "ws://localhost:3001",
        api_token: Optional[str] = None,
    ):
        self.base_url = base_url.replace("http", "ws")
        self.api_token = api_token
        self._ws: Optional[websockets.WebSocketClientProtocol] = None
        self._event_callback: Optional[Callable[[SessionEvent], Awaitable[None]]] = None

    async def connect(
        self,
        on_event: Callable[[SessionEvent], Awaitable[None]],
    ) -> None:
        self._event_callback = on_event

        url = self.base_url
        if self.api_token:
            url = f"{url}?token={self.api_token}"

        try:
            self._ws = await websockets.connect(url)
            asyncio.create_task(self._receive_loop())
        except Exception as e:
            raise TextQuestError(f"WebSocket connection failed: {e}")

    async def _receive_loop(self) -> None:
        if not self._ws:
            return

        try:
            async for message in self._ws:
                try:
                    data = json.loads(message)
                    event = self._parse_event(data)
                    if self._event_callback:
                        await self._event_callback(event)
                except json.JSONDecodeError:
                    pass
        except websockets.ConnectionClosed:
            pass
        finally:
            self._ws = None

    def _parse_event(self, data: Dict[str, Any]) -> SessionEvent:
        event_type = data.get("type", "unknown")
        session_type = SessionEventType(event_type)

        return SessionEvent(type=session_type, data=data)

    async def send_command(self, command: str, target: Optional[str] = None) -> None:
        if not self._ws:
            raise TextQuestError("WebSocket not connected")

        message = {"type": "command", "command": command}
        if target:
            message["target"] = target

        await self._ws.send(json.dumps(message))

    async def disconnect(self) -> None:
        if self._ws:
            await self._ws.close()
            self._ws = None

File: textquest-client/typescript/src/websocket.ts
import type { SessionEvent, ClientConfig } from './types.js';
import { TextQuestClientError } from './types.js';

type EventCallback = (event: SessionEvent) => void;

export class TextQuestWebSocket {
  private ws: WebSocket | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelayMs = 1000;
  private eventCallback?: EventCallback;
  private url: string;

  constructor(private config: ClientConfig) {
    this.url = config.baseUrl.replace('http', 'ws').replace(/\/$/, '') + '/ws';
  }

  private buildAuthUrl(): string {
    if (this.config.apiToken) {
      return `${this.url}?token=${encodeURIComponent(this.config.apiToken)}`;
    }
    return this.url;
  }

  async connect(onEvent: EventCallback): Promise<void> {
    return new Promise((resolve, reject) => {
      this.eventCallback = onEvent;

      try {
        this.ws = new WebSocket(this.buildAuthUrl());

        this.ws.onopen = () => {
          console.log('WebSocket connected');
          this.reconnectAttempts = 0;
          resolve();
        };

        this.ws.onmessage = (event) => {
          try {
            const data = JSON.parse(event.data);
            const sessionEvent: SessionEvent = this.parseEvent(data);
            if (this.eventCallback) {
              this.eventCallback(sessionEvent);
            }
          } catch (e) {
            console.error('Failed to parse WebSocket message:', e);
          }
        };

        this.ws.onclose = () => {
          console.log('WebSocket disconnected');
          this.attemptReconnect();
        };

        this.ws.onerror = (error) => {
          console.error('WebSocket error:', error);
          reject(new TextQuestClientError('WebSocket connection failed'));
        };
      } catch (e) {
        reject(e);
      }
    });
  }

  private parseEvent(data: unknown): SessionEvent {
    const obj = data as Record<string, unknown>;
    const type = (obj.type as string) || 'unknown';

    switch (type) {
      case 'session':
      case 'session_update':
        return {
          type: 'session_update',
          session: obj.session as any,
        };
      case 'command':
        return {
          type: 'command',
          command: obj.command as string,
          target: obj.target as string | undefined,
        };
      default:
        return {
          type: 'error',
          message: `Unknown event type: ${type}`,
        };
    }
  }

  private async attemptReconnect(): Promise<void> {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.error('Max reconnect attempts reached');
      return;
    }

    this.reconnectAttempts++;
    const delay = this.reconnectDelayMs * Math.pow(2, this.reconnectAttempts - 1);
    console.log(`Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts})...`);

    await new Promise((resolve) => setTimeout(resolve, delay));

    if (this.eventCallback) {
      try {
        await this.connect(this.eventCallback);
      } catch (e) {
        console.error('Reconnect failed:', e);
      }
    }
  }

  send(command: string, target?: string): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new TextQuestClientError('WebSocket not connected');
    }

    const message = JSON.stringify({
      type: 'command',
      command,
      target,
    });
    this.ws.send(message);
  }

  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}